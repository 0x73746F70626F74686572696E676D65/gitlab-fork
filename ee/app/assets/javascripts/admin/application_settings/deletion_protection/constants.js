import { __, s__ } from '~/locale';

export const I18N_DELETION_PROTECTION = {
  heading: s__('AdminSettings|Deletion protection'),
  helpText: s__(
    'AdminSettings|Retention period that deleted groups and projects will remain restorable. Personal projects are always deleted immediately. Some groups can opt-out their projects.',
  ),
  learnMore: __('Learn more.'),
  keepDeleted: s__('AdminSettings|Keep deleted'),
  deleteImmediately: s__('AdminSettings|None, delete immediately'),
  for: __('for'),
  days: __('days'),
  groupsOnly: __('groups only'),
  groupsAndProjects: __('groups and projects'),
};
