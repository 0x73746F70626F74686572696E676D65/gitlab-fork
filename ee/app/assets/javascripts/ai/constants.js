import { s__, sprintf } from '~/locale';

export const i18n = {
  GENIE_TOOLTIP: s__('AI|What does the selected code mean?'),
  GENIE_NO_CONTAINER_ERROR: s__("AI|The container element wasn't found, stopping AI Genie."),
  GENIE_CHAT_TITLE: s__('AI|Code Explanation'),
  GENIE_CHAT_CLOSE_LABEL: s__('AI|Close the Code Explanation'),
  GENIE_CHAT_LEGAL_NOTICE: sprintf(
    s__(
      'AI|You are not allowed to copy any part of this output into issues, comments, GitLab source code, commit messages, merge requests or any other user interface in the %{gitlabOrg} or %{gitlabCom} groups.',
    ),
    { gitlabOrg: '<code>/gitlab-org</code>', gitlabCom: '<code>/gitlab-com</code>' },
    false,
  ),
  GENIE_CHAT_LEGAL_GENERATED_BY_AI: s__('AI|Responses generated by AI'),
  REQUEST_ERROR: s__('AI|Something went wrong. Please try again later'),
  EXPERIMENT_BADGE: s__('AI|Experiment'),
  FEEDBACK_LINK: s__('AI|Give feedback on code explanation'),
};

export const AI_GENIE_DEBOUNCE = 300;

export const FEEDBACK_LINK_URL = 'https://gitlab.com/gitlab-org/gitlab/-/issues/407285';
