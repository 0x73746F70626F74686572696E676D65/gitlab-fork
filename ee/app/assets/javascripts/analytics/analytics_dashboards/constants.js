import { s__ } from '~/locale';

export const FEATURE_PRODUCT_ANALYTICS = 'productAnalytics';

export const PRODUCT_ANALYTICS_FEATURE_DASHBOARDS = ['audience', 'behavior'];

export const EVENTS_TYPES = ['pageViews', 'linkClickEvents', 'events'];

export function isRestrictedToEventType(eventType) {
  return EVENTS_TYPES.includes(eventType);
}

export const PANEL_VISUALIZATION_HEIGHT = '600px';

export const PANEL_DISPLAY_TYPES = {
  DATA: 'data',
  VISUALIZATION: 'visualization',
  CODE: 'code',
};

export const PANEL_DISPLAY_TYPE_ITEMS = [
  {
    type: PANEL_DISPLAY_TYPES.DATA,
    icon: 'table',
    title: s__('Analytics|Data'),
  },
  {
    type: PANEL_DISPLAY_TYPES.VISUALIZATION,
    icon: 'chart',
    title: s__('Analytics|Visualization'),
  },
  {
    type: PANEL_DISPLAY_TYPES.CODE,
    icon: 'code',
    title: s__('Analytics|Code'),
  },
];

export const MEASURE_COLOR = '#00b140';
export const DIMENSION_COLOR = '#c3e6cd';

export const EVENTS_TABLE_NAME = 'TrackedEvents';
export const SESSIONS_TABLE_NAME = 'Sessions';
export const RETURNING_USERS_TABLE_NAME = 'ReturningUsers';

export const TRACKED_EVENTS_KEY = 'trackedevents';

export const ANALYTICS_FIELD_CATEGORIES = [
  {
    name: s__('Analytics|Pages'),
    category: 'pages',
  },
  {
    name: s__('Analytics|Users'),
    category: 'users',
  },
  {
    name: s__('Analytics|Link clicks'),
    category: 'linkClicks',
  },
  {
    name: s__('Analytics|Custom events'),
    category: 'custom',
  },
];

export const ANALYTICS_FIELDS = [
  {
    name: s__('Analytics|URL'),
    category: 'pages',
    dbField: 'pageUrl',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Page Path'),
    category: 'pages',
    dbField: 'pageUrlpath',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Page Title'),
    category: 'pages',
    dbField: 'pageTitle',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Page Language'),
    category: 'pages',
    dbField: 'documentLanguage',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Host'),
    category: 'pages',
    dbField: 'pageUrlhosts',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Referer'),
    category: 'users',
    dbField: 'pageReferrer',
    icon: 'user',
  },
  {
    name: s__('Analytics|Language'),
    category: 'users',
    dbField: 'browserLanguage',
    icon: 'user',
  },
  {
    name: s__('Analytics|Viewport'),
    category: 'users',
    dbField: 'viewportSize',
    icon: 'user',
  },
  {
    name: s__('Analytics|Browser Family'),
    category: 'users',
    dbField: 'agentName',
    icon: 'user',
  },
  {
    name: s__('Analytics|Browser'),
    category: 'users',
    dbField: ['agentName', 'agentVersion'],
    icon: 'user',
  },
  {
    name: s__('Analytics|OS'),
    category: 'users',
    dbField: 'osName',
    icon: 'user',
  },
  {
    name: s__('Analytics|OS Version'),
    category: 'users',
    dbField: ['osName', 'osVersion'],
    icon: 'user',
  },
  {
    name: s__('Analytics|User Id'),
    category: 'users',
    dbField: 'userId',
    icon: 'user',
  },
  {
    name: s__('Analytics|Target URL'),
    category: 'linkClicks',
    dbField: ['targetUrl'],
    icon: 'link',
  },
  {
    name: s__('Analytics|Element ID'),
    category: 'linkClicks',
    dbField: ['elementId'],
    icon: 'link',
  },
  {
    name: s__('Analytics|Event Name'),
    category: 'custom',
    dbField: 'customEventName',
    icon: 'documents',
  },
  {
    name: s__('Analytics|Event Props'),
    category: 'custom',
    dbField: 'customEventProps',
    icon: 'documents',
  },
  {
    name: s__('Analytics|User Props'),
    category: 'custom',
    dbField: 'customUserProps',
    icon: 'user',
  },
];

export const NEW_DASHBOARD = () => ({
  title: '',
  description: '',
  panels: [],
  userDefined: true,
});

export const FILE_ALREADY_EXISTS_SERVER_RESPONSE = 'A file with this name already exists';

export const EVENT_LABEL_CREATED_DASHBOARD = 'user_created_custom_dashboard';
export const EVENT_LABEL_EDITED_DASHBOARD = 'user_edited_custom_dashboard';
export const EVENT_LABEL_VIEWED_DASHBOARD_DESIGNER = 'user_viewed_dashboard_designer';
export const EVENT_LABEL_VIEWED_CUSTOM_DASHBOARD = 'user_viewed_custom_dashboard';

export const EVENT_LABEL_USER_VIEWED_VISUALIZATION_DESIGNER = 'user_viewed_visualization_designer';
export const EVENT_LABEL_USER_CREATED_CUSTOM_VISUALIZATION = 'user_created_custom_visualization';

export const EVENT_LABEL_EXCLUDE_ANONYMISED_USERS = 'exclude_anonymised_users';
