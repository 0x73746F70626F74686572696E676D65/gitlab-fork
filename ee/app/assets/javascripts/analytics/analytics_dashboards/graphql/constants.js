export const TYPENAME_PRODUCT_ANALYTICS_DASHBOARD = 'CustomizableDashboards';
export const TYPENAME_PRODUCT_ANALYTICS_DASHBOARD_CONNECTION = 'CustomizableDashboardsConnection';
