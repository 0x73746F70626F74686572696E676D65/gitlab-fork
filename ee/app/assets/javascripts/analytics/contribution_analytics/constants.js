import { __, s__ } from '~/locale';

export const CHART_HEIGHT = 350;
export const INNER_CHART_HEIGHT = 200;
export const CHART_X_AXIS_ROTATE = 45;
export const CHART_X_AXIS_NAME_TOP_PADDING = 55;

export const LEGACY_TABLE_COLUMNS = [
  { name: 'fullname', text: __('Name') },
  { name: 'push', text: __('Pushed') },
  { name: 'issuesCreated', text: __('Opened issues') },
  { name: 'issuesClosed', text: __('Closed issues') },
  { name: 'mergeRequestsCreated', text: __('Opened MRs') },
  { name: 'mergeRequestsApproved', text: __('Approved MRs') },
  { name: 'mergeRequestsMerged', text: __('Merged MRs') },
  { name: 'mergeRequestsClosed', text: __('Closed MRs') },
  { name: 'totalEvents', text: __('Total Contributions') },
];

export const TABLE_COLUMNS = [
  { key: 'user', label: s__('ContributionAnalytics|Name') },
  { key: 'repoPushed', label: s__('ContributionAnalytics|Pushed') },
  { key: 'issuesCreated', label: s__('ContributionAnalytics|Opened issues') },
  { key: 'issuesClosed', label: s__('ContributionAnalytics|Closed issues') },
  { key: 'mergeRequestsCreated', label: s__('ContributionAnalytics|Opened MRs') },
  { key: 'mergeRequestsApproved', label: s__('ContributionAnalytics|Approved MRs') },
  { key: 'mergeRequestsMerged', label: s__('ContributionAnalytics|Merged MRs') },
  { key: 'mergeRequestsClosed', label: s__('ContributionAnalytics|Closed MRs') },
  { key: 'totalEvents', label: s__('ContributionAnalytics|Total Contributions') },
].map((col) => ({
  ...col,
  sortable: true,
  thClass: 'gl-vertical-align-middle!',
}));
