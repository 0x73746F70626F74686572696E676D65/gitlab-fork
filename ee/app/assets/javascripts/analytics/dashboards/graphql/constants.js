export const BUCKETING_INTERVAL_ALL = 'ALL';
