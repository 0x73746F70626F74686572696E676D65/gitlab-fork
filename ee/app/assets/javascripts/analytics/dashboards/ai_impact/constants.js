import { pick } from 'lodash';
import { s__ } from '~/locale';
import {
  FLOW_METRICS,
  DORA_METRICS,
  VULNERABILITY_METRICS,
  AI_METRICS,
} from '~/analytics/shared/constants';

import { UNITS, TABLE_METRICS as VSD_TABLE_METRICS } from '../constants';

export const SUPPORTED_FLOW_METRICS = [FLOW_METRICS.CYCLE_TIME, FLOW_METRICS.LEAD_TIME];

export const SUPPORTED_DORA_METRICS = [
  DORA_METRICS.DEPLOYMENT_FREQUENCY,
  DORA_METRICS.CHANGE_FAILURE_RATE,
];

export const SUPPORTED_VULNERABILITY_METRICS = [VULNERABILITY_METRICS.CRITICAL];

export const SUPPORTED_AI_METRICS = [AI_METRICS.CODE_SUGGESTIONS_USAGE_RATE];

export const TABLE_METRICS = {
  ...pick(VSD_TABLE_METRICS, [
    ...SUPPORTED_FLOW_METRICS,
    ...SUPPORTED_DORA_METRICS,
    ...SUPPORTED_VULNERABILITY_METRICS,
  ]),
  [AI_METRICS.CODE_SUGGESTIONS_USAGE_RATE]: {
    label: s__('DORA4Metrics|Code Suggestions usage'),
    units: UNITS.PERCENT,
  },
};
