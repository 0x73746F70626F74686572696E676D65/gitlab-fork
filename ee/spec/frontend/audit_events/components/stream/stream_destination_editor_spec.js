import VueApollo from 'vue-apollo';
import { createLocalVue } from '@vue/test-utils';
import { GlButton, GlFormCheckbox, GlForm, GlTableLite } from '@gitlab/ui';
import * as Sentry from '@sentry/browser';
import { sprintf } from '~/locale';
import createMockApollo from 'helpers/mock_apollo_helper';
import waitForPromises from 'helpers/wait_for_promises';
import {
  shallowMountExtended,
  mountExtended,
  extendedWrapper,
} from 'helpers/vue_test_utils_helper';
import externalAuditEventDestinationCreate from 'ee/audit_events/graphql/create_external_destination.mutation.graphql';
import externalAuditEventDestinationHeaderCreate from 'ee/audit_events/graphql/create_external_destination_header.mutation.graphql';
import deleteExternalDestination from 'ee/audit_events/graphql/delete_external_destination.mutation.graphql';
import StreamDestinationEditor from 'ee/audit_events/components/stream/stream_destination_editor.vue';
import { AUDIT_STREAMS_NETWORK_ERRORS, ADD_STREAM_EDITOR_I18N } from 'ee/audit_events/constants';
import {
  destinationCreateMutationPopulator,
  destinationDeleteMutationPopulator,
  destinationHeaderCreateMutationPopulator,
  groupPath,
} from '../../mock_data';

const localVue = createLocalVue();
localVue.use(VueApollo);

describe('StreamDestinationEditor', () => {
  let wrapper;

  const maxHeaders = 3;

  const createComponent = (
    mountFn = shallowMountExtended,
    provide = {},
    apolloHandlers = [
      [
        externalAuditEventDestinationCreate,
        jest.fn().mockResolvedValue(destinationCreateMutationPopulator()),
      ],
    ],
  ) => {
    const mockApollo = createMockApollo(apolloHandlers);
    wrapper = mountFn(StreamDestinationEditor, {
      provide: {
        groupPath,
        showStreamsHeaders: false,
        maxHeaders,
        ...provide,
      },
      apolloProvider: mockApollo,
      localVue,
    });
  };

  const findWarningMessage = () => wrapper.findByTestId('data-warning');
  const findAlertErrors = () => wrapper.findAllByTestId('alert-errors');
  const findDestinationForm = () => wrapper.findComponent(GlForm);
  const findHeadersTable = () => wrapper.findComponent(GlTableLite);
  const findMaximumHeadersText = () => wrapper.findByTestId('maximum-headers').text();
  const findAddBtn = () => wrapper.findByTestId('stream-destination-add-button');
  const findCancelBtn = () => wrapper.findByTestId('stream-destination-cancel-button');

  const findDestinationUrlFormGroup = () => wrapper.findByTestId('destination-url-form-group');
  const findDestinationUrl = () => wrapper.findByTestId('destination-url');

  const findHeadersRows = () => findHeadersTable().find('tbody').findAll('tr');
  const findHeadersHeaderCell = (tdIdx) =>
    findHeadersTable().find('thead tr').findAll('th').at(tdIdx);
  const findHeadersCell = (trIdx, tdIdx) => findHeadersRows().at(trIdx).findAll('td').at(tdIdx);
  const findHeaderCheckbox = (trIdx) => findHeadersRows().at(trIdx).findComponent(GlFormCheckbox);
  const findHeaderDeleteBtn = (trIdx) => findHeadersRows().at(trIdx).findComponent(GlButton);
  const findHeaderNameInput = (trIdx) =>
    extendedWrapper(findHeadersRows().at(trIdx)).findByTestId('header-name-input');
  const findHeaderValueInput = (trIdx) =>
    extendedWrapper(findHeadersRows().at(trIdx)).findByTestId('header-value-input');

  const setHeaderNameInput = (trIdx, name) => findHeaderNameInput(trIdx).setValue(name);
  const setHeaderValueInput = (trIdx, value) => findHeaderValueInput(trIdx).setValue(value);

  const setHeadersRowData = async (trIdx, { name, value }) => {
    await setHeaderNameInput(trIdx, name);
    await setHeaderValueInput(trIdx, value);
  };

  afterEach(() => {
    wrapper.destroy();
  });

  describe('when initialized', () => {
    describe('destinations URL', () => {
      beforeEach(() => {
        createComponent();
      });

      it('should render the destinations warning', () => {
        expect(findWarningMessage().props('title')).toBe(ADD_STREAM_EDITOR_I18N.WARNING_TITLE);
        expect(findWarningMessage().text()).toBe(ADD_STREAM_EDITOR_I18N.WARNING_CONTENT);
      });

      it('should render the destination URL input', () => {
        expect(findDestinationUrlFormGroup().exists()).toBe(true);
        expect(findDestinationUrl().attributes('placeholder')).toBe(
          ADD_STREAM_EDITOR_I18N.DESTINATION_URL_PLACEHOLDER,
        );
      });
    });

    describe('HTTP headers', () => {
      beforeEach(() => {
        createComponent(mountExtended, { showStreamsHeaders: true });
      });

      it('should render the table', () => {
        expect(findHeadersRows()).toHaveLength(1);

        expect(findHeadersHeaderCell(0).text()).toBe(
          ADD_STREAM_EDITOR_I18N.TABLE_COLUMN_NAME_LABEL,
        );
        expect(findHeadersHeaderCell(1).text()).toBe(
          ADD_STREAM_EDITOR_I18N.TABLE_COLUMN_VALUE_LABEL,
        );
        expect(findHeadersHeaderCell(2).text()).toBe(
          ADD_STREAM_EDITOR_I18N.TABLE_COLUMN_ACTIVE_LABEL,
        );
        expect(findHeadersHeaderCell(3).text()).toBe('');

        expect(findHeaderNameInput(0).attributes('placeholder')).toBe(
          ADD_STREAM_EDITOR_I18N.HEADER_INPUT_PLACEHOLDER,
        );
        expect(findHeaderValueInput(0).attributes('placeholder')).toBe(
          ADD_STREAM_EDITOR_I18N.VALUE_INPUT_PLACEHOLDER,
        );
        expect(findHeaderCheckbox(0).find('input').attributes('disabled')).toBe('disabled');
        expect(findHeaderDeleteBtn(0).exists()).toBe(true);
      });
    });
  });

  describe('add destination event without headers', () => {
    it('should emit add event after destination added', async () => {
      createComponent(shallowMountExtended, {}, [
        [
          externalAuditEventDestinationCreate,
          jest.fn().mockResolvedValue(destinationCreateMutationPopulator()),
        ],
      ]);

      findDestinationUrl().vm.$emit('input', 'https://example.test');
      findDestinationForm().vm.$emit('submit', { preventDefault: () => {} });
      await waitForPromises();

      expect(findAlertErrors()).toHaveLength(0);
      expect(wrapper.emitted('added')).toBeDefined();
    });

    it('should not emit add destination event and reports error when server returns error', async () => {
      const errorMsg = 'Destination hosts limit exceeded';
      createComponent(shallowMountExtended, {}, [
        [
          externalAuditEventDestinationCreate,
          jest.fn().mockResolvedValue(destinationCreateMutationPopulator([errorMsg])),
        ],
      ]);

      findDestinationUrl().vm.$emit('input', 'https://example.test');
      findDestinationForm().vm.$emit('submit', { preventDefault: () => {} });
      await waitForPromises();

      expect(findAlertErrors()).toHaveLength(1);
      expect(findAlertErrors().at(0).text()).toBe(errorMsg);
      expect(wrapper.emitted('added')).not.toBeDefined();
    });

    it('should not emit add destination event and reports error when network error occurs', async () => {
      const sentryError = new Error('Network error');
      const sentryCaptureExceptionSpy = jest.spyOn(Sentry, 'captureException');
      createComponent(shallowMountExtended, {}, [
        [externalAuditEventDestinationCreate, jest.fn().mockRejectedValue(sentryError)],
      ]);

      findDestinationUrl().vm.$emit('input', 'https://example.test');
      findDestinationForm().vm.$emit('submit', { preventDefault: () => {} });
      await waitForPromises();

      expect(findAlertErrors()).toHaveLength(1);
      expect(findAlertErrors().at(0).text()).toBe(AUDIT_STREAMS_NETWORK_ERRORS.CREATING_ERROR);
      expect(sentryCaptureExceptionSpy).toHaveBeenCalledWith(sentryError);
      expect(wrapper.emitted('added')).not.toBeDefined();
    });
  });

  describe('add destination event with headers', () => {
    it('should emit add event after destination and headers are added', async () => {
      createComponent(mountExtended, { showStreamsHeaders: true }, [
        [
          externalAuditEventDestinationCreate,
          jest.fn().mockResolvedValue(destinationCreateMutationPopulator()),
        ],
        [
          externalAuditEventDestinationHeaderCreate,
          jest.fn().mockResolvedValue(destinationHeaderCreateMutationPopulator()),
        ],
      ]);

      findDestinationUrl().vm.$emit('input', 'https://example.test');
      await setHeadersRowData(0, { name: 'row header', value: 'row value' });
      await setHeadersRowData(1, { name: 'row header 1', value: 'row value 1' });
      findDestinationForm().vm.$emit('submit', { preventDefault: () => {} });
      await waitForPromises();

      expect(findAlertErrors()).toHaveLength(0);
      expect(wrapper.emitted('added')).toBeDefined();
    });

    it('should ignore empty headers and emit add event after destination and headers are added', async () => {
      const headerCreateSpy = jest
        .fn()
        .mockResolvedValue(destinationHeaderCreateMutationPopulator());

      createComponent(mountExtended, { showStreamsHeaders: true }, [
        [
          externalAuditEventDestinationCreate,
          jest.fn().mockResolvedValue(destinationCreateMutationPopulator()),
        ],
        [externalAuditEventDestinationHeaderCreate, headerCreateSpy],
      ]);

      findDestinationUrl().vm.$emit('input', 'https://example.test');
      await setHeadersRowData(0, { name: 'row header', value: 'row value' });
      await setHeadersRowData(1, { name: '', value: '' });
      findDestinationForm().vm.$emit('submit', { preventDefault: () => {} });
      await waitForPromises();

      expect(findAlertErrors()).toHaveLength(0);
      expect(headerCreateSpy).toHaveBeenCalledTimes(1);
      expect(headerCreateSpy).toHaveBeenCalledWith({
        destinationId: 'test-create-id',
        key: 'row header',
        value: 'row value',
      });
      expect(wrapper.emitted('added')).toBeDefined();
    });

    it('should not emit add destination event and reports error when server returns error while adding headers', async () => {
      const errorMsg = 'Destination hosts limit exceeded';
      createComponent(mountExtended, { showStreamsHeaders: true }, [
        [
          externalAuditEventDestinationCreate,
          jest.fn().mockResolvedValue(destinationCreateMutationPopulator()),
        ],
        [
          externalAuditEventDestinationHeaderCreate,
          jest
            .fn()
            .mockResolvedValueOnce(destinationHeaderCreateMutationPopulator())
            .mockResolvedValue(destinationHeaderCreateMutationPopulator([errorMsg])),
        ],
        [
          deleteExternalDestination,
          jest.fn().mockResolvedValue(destinationDeleteMutationPopulator()),
        ],
      ]);

      findDestinationUrl().vm.$emit('input', 'https://example.test');
      await setHeadersRowData(0, { name: 'row header', value: 'row value' });
      await setHeadersRowData(1, { name: 'row header 1', value: 'row value 1' });
      findDestinationForm().vm.$emit('submit', { preventDefault: () => {} });
      await waitForPromises();

      expect(findAlertErrors()).toHaveLength(1);
      expect(findAlertErrors().at(0).text()).toBe(errorMsg);
      expect(wrapper.emitted('added')).not.toBeDefined();
    });

    it('should not emit add destination event and reports error when network error occurs while adding headers', async () => {
      const sentryError = new Error('Network error');
      const sentryCaptureExceptionSpy = jest.spyOn(Sentry, 'captureException');
      createComponent(mountExtended, { showStreamsHeaders: true }, [
        [
          externalAuditEventDestinationCreate,
          jest.fn().mockResolvedValue(destinationCreateMutationPopulator()),
        ],
        [
          externalAuditEventDestinationHeaderCreate,
          jest
            .fn()
            .mockResolvedValueOnce(destinationHeaderCreateMutationPopulator())
            .mockRejectedValue(sentryError),
        ],
        [
          deleteExternalDestination,
          jest.fn().mockResolvedValue(destinationDeleteMutationPopulator()),
        ],
      ]);

      findDestinationUrl().vm.$emit('input', 'https://example.test');
      await setHeadersRowData(0, { name: 'row header', value: 'row value' });
      await setHeadersRowData(1, { name: 'row header 1', value: 'row value 1' });
      findDestinationForm().vm.$emit('submit', { preventDefault: () => {} });
      await waitForPromises();

      expect(findAlertErrors()).toHaveLength(1);
      expect(findAlertErrors().at(0).text()).toBe(AUDIT_STREAMS_NETWORK_ERRORS.CREATING_ERROR);
      expect(sentryCaptureExceptionSpy).toHaveBeenCalledWith(sentryError);
      expect(wrapper.emitted('added')).not.toBeDefined();
    });
  });

  describe('cancel event', () => {
    beforeEach(() => {
      createComponent();
    });

    it('should emit cancel event correctly', () => {
      findCancelBtn().vm.$emit('click');

      expect(wrapper.emitted('cancel')).toBeDefined();
    });
  });

  describe('HTTP headers table', () => {
    beforeEach(() => {
      createComponent(mountExtended, { showStreamsHeaders: true });
    });

    it.each`
      name     | value    | rowCount
      ${'abc'} | ${''}    | ${1}
      ${''}    | ${'abc'} | ${1}
      ${'abc'} | ${'abc'} | ${2}
    `(
      'should add a new blank row only when both the name and value are filled',
      async ({ name, value, rowCount }) => {
        await setHeadersRowData(0, { name, value });

        expect(findHeadersRows()).toHaveLength(rowCount);
      },
    );

    it.each`
      name     | value    | disabled
      ${'abc'} | ${''}    | ${true}
      ${''}    | ${'abc'} | ${true}
      ${'abc'} | ${'abc'} | ${false}
    `(
      'should enable the add button only when both the name and value are filled',
      async ({ name, value, disabled }) => {
        await findDestinationUrl().setValue('https://example.test');
        await setHeadersRowData(0, { name, value });

        expect(findAddBtn().props('disabled')).toBe(disabled);
      },
    );

    it('should not add a new blank row if the header value is a duplicate', async () => {
      await findDestinationUrl().setValue('https://example.test');
      await setHeadersRowData(0, { name: 'row header', value: 'row value' });

      expect(findHeadersRows()).toHaveLength(2);

      await setHeadersRowData(1, { name: 'row header', value: 'row value' });

      expect(findHeadersRows()).toHaveLength(2);
      expect(findHeaderNameInput(1).classes()).toContain('is-invalid');
      expect(findHeadersCell(1, 0).text()).toContain(
        ADD_STREAM_EDITOR_I18N.HEADER_INPUT_DUPLICATE_ERROR,
      );
      expect(findAddBtn().props('disabled')).toBe(true);
    });

    it('should add a new blank row once a duplicate value is changed', async () => {
      await findDestinationUrl().setValue('https://example.test');
      await setHeadersRowData(0, { name: 'row header', value: 'row value' });
      await setHeadersRowData(1, { name: 'row header', value: 'row value' });

      expect(findHeadersRows()).toHaveLength(2);
      expect(findHeaderNameInput(1).classes()).toContain('is-invalid');
      expect(findAddBtn().props('disabled')).toBe(true);

      await setHeadersRowData(1, { name: 'row header 2', value: 'row value' });

      expect(findHeaderNameInput(1).classes()).toContain('is-valid');
      expect(findAddBtn().props('disabled')).toBe(false);
      expect(findHeadersRows()).toHaveLength(3);
    });

    it('should delete a row when the delete button is clicked', async () => {
      await setHeadersRowData(0, { name: 'row header', value: 'row value' });
      await setHeadersRowData(1, { name: 'row header 2', value: 'row value 2' });

      expect(findHeadersRows()).toHaveLength(3);

      await findHeaderDeleteBtn(1).trigger('click');

      expect(findHeadersRows()).toHaveLength(2);
      expect(findHeaderNameInput(0).element.value).toBe('row header');
      expect(findHeaderValueInput(0).element.value).toBe('row value');
      expect(findHeaderNameInput(1).element.value).toBe('');
      expect(findHeaderValueInput(1).element.value).toBe('');
    });

    it('should add a blank row if the last row was deleted', async () => {
      await setHeadersRowData(0, { name: 'row header', value: '' });

      expect(findHeadersRows()).toHaveLength(1);

      await findHeaderDeleteBtn(0).trigger('click');

      expect(findHeadersRows()).toHaveLength(1);
      expect(findHeaderNameInput(0).element.value).toBe('');
      expect(findHeaderValueInput(0).element.value).toBe('');
    });

    it('should show the maximum number of rows message when the maximum is reached', async () => {
      for (let i = 0; i < maxHeaders; i += 1) {
        // This should be done synchronously because each new input will trigger a new one up to the maximum
        // eslint-disable-next-line no-await-in-loop
        await setHeadersRowData(i, { name: `row header ${i}`, value: `row value ${i}` });
      }

      expect(findHeadersRows()).toHaveLength(maxHeaders);
      expect(findMaximumHeadersText()).toMatchInterpolatedText(
        sprintf(ADD_STREAM_EDITOR_I18N.MAXIMUM_HEADERS_TEXT, { number: maxHeaders }),
      );
    });
  });
});
