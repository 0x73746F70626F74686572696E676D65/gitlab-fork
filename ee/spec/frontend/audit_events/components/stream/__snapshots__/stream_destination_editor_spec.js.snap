// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamDestinationEditor when initialized should render correctly 1`] = `
<div
  class="gl-p-4 gl-bg-white gl-border gl-rounded-base"
>
  <div
    class="gl-mb-5 gl-alert gl-alert-not-dismissible gl-alert-warning"
  >
    <svg
      aria-hidden="true"
      class="gl-icon s16 gl-alert-icon"
      data-testid="warning-icon"
      role="img"
    >
      <use
        href="#warning"
      />
    </svg>
     
    <div
      class="gl-alert-content"
      role="alert"
    >
      <h2
        class="gl-alert-title"
      >
        Destinations receive all audit event data
      </h2>
       
      <div
        class="gl-alert-body"
      >
        
    This could include sensitive information. Make sure you trust the destination endpoint.
  
      </div>
       
      <!---->
    </div>
     
    <!---->
  </div>
   
  <form
    class=""
  >
    <fieldset
      aria-describedby=""
      class="form-group gl-form-group"
      id="__BVID__8"
    >
      <legend
        class="bv-no-focus-ring col-form-label pt-0 col-form-label"
        id="__BVID__8__BV_label_"
        tabindex="-1"
      >
        
      Destination URL
      
        <!---->
         
        <!---->
      </legend>
      <div
        aria-labelledby="__BVID__8__BV_label_"
        class="bv-no-focus-ring"
        role="group"
        tabindex="-1"
      >
        <input
          class="gl-form-input form-control"
          id="__BVID__10"
          placeholder="https://api.gitlab.com"
          type="text"
        />
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
     
    <div
      class="gl-display-flex"
    >
      <button
        class="btn gl-mr-3 btn-info btn-md disabled gl-button"
        data-testid="stream-destination-add-button"
        disabled="disabled"
        name="Add external stream destination"
        type="submit"
      >
        <!---->
         
        <!---->
          
        <span
          class="gl-button-text"
        >
          Add
        </span>
      </button>
       
      <button
        class="btn btn-default btn-md gl-button"
        data-testid="stream-destination-cancel-button"
        name="AuditStreams|Cancel editing"
        type="button"
      >
        <!---->
         
        <!---->
          
        <span
          class="gl-button-text"
        >
          Cancel
        </span>
      </button>
    </div>
  </form>
</div>
`;
